src/java/org/apache/cassandra/index/sai/memory/MemtableRangeIterator.java

```diff
@@ -88,33 +86,17 @@ private static PrimaryKey maxKey(Memtable memtable, PrimaryKey.Factory factory)
     }
 
     @Override
-    protected void performSkipTo(PrimaryKey nextKey)
+    protected void performSkipTo(Token nextToken)
     {
-        PartitionPosition start = nextKey.partitionKey() != null
-                                  ? nextKey.partitionKey()
-                                  : nextKey.token().minKeyBound();
-        if (!keyRange.right.isMinimum() && start.compareTo(keyRange.right) > 0)
+        do
         {
-            partitionIterator = EmptyIterators.unfilteredPartition(memtable.metadata());
-            rowIterator = null;
-            return;
-        }
-
-        AbstractBounds<PartitionPosition> partitionBounds = AbstractBounds.bounds(start, true, keyRange.right, true);
-        DataRange dataRange = new DataRange(partitionBounds, new ClusteringIndexSliceFilter(Slices.ALL, false));
-        FileUtils.closeQuietly(partitionIterator);
-        partitionIterator = memtable.makePartitionIterator(columns, dataRange);
-        if (partitionIterator.hasNext())
-        {
-            this.rowIterator = partitionIterator.next();
-            if (!nextKey.hasEmptyClustering() && rowIterator.partitionKey().equals(nextKey.partitionKey()))
-            {
-                Slice slice = Slice.make(nextKey.clustering(), Clustering.EMPTY);
-                Slices slices = Slices.with(memtable.metadata().comparator, slice);
-                FileUtils.closeQuietly(rowIterator);
-                rowIterator = memtable.getPartition(nextKey.partitionKey()).unfilteredIterator(columns, slices, false);
-            }
-        }
+            if (!rowIterator.hasNext())
+                continue;
+            rowIterator = partitionIterator.next();
+            if (nextToken.compareTo(rowIterator.partitionKey().getToken()) <= 0)
+                return;
+            FileUtils.closeQuietly(rowIterator);
+        } while (partitionIterator.hasNext());
```

Comment:
@pkolaczk - I wasn't sure about this implementation's efficiency. I went with it because of its simplicity. Let me know if this is a problem.